### ISSUE TITLE
jaxlibWithCuda11 fails to build

### ISSUE THREAD

**Issue by baduhai** [nixvital/ml-pkgs#4]
I've tried pinning nixpkgs to the revision in this repo, as well as to 2d10e73416ec1449ef74aeac7faf2cf8c556ff5a, which is the same revision as when jaxlib was no longer marked as "fails to build" on this repo. I'm currently building against the latest stable nixpkgs, but I doubt it's going to work, since I've tried building it without this overlay, and no luck. This may be an upstream issue.

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/0sd34630xb88aca480v5wr8920s5pgnc-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
patching script interpreter paths in ../output/external/org_tensorflow/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl
../output/external/org_tensorflow/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl: interpreter directive changed from "#!/usr/bin/env python" to "/nix/store/zdba9frlxj2ba8ca095win3nphsiiqhb-python3-3.10.8/bin/python"
Extracting Bazel installation...
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=0
INFO: Reading rc options for 'build' from /build/source/.bazelrc:
  Inherited 'common' options: --experimental_repo_remote_exec
INFO: Reading rc options for 'build' from /nix/store/xn6ql92hqwqh1yvvwnv6gmhi52hfpbnb-bazel-rc:
  'build' options: --distdir=/nix/store/hcn6rxvvzig59nh7k3rmng8j809lv3ap-bazel-deps --extra_toolchains=@bazel_tools//tools/jdk:nonprebuilt_toolchain_definition --tool_java_runtime_version=local_jdk_11 --java_runtime_version=local_jdk_11
INFO: Reading rc options for 'build' from /build/source/.bazelrc:
  'build' options: --apple_platform_type=macos --macos_minimum_os=10.14 --announce_rc --define open_source_build=true --spawn_strategy=standalone --enable_platform_specific_config --experimental_cc_shared_library --define=no_aws_support=true --define=no_gcp_support=true --define=no_hdfs_support=true --define=no_kafka_support=true --define=no_ignite_support=true --define=grpc_n>
INFO: Reading rc options for 'build' from /build/source/.jax_configure.bazelrc:
  'build' options: --strategy=Genrule=standalone --repo_env PYTHON_BIN_PATH=/nix/store/zdba9frlxj2ba8ca095win3nphsiiqhb-python3-3.10.8/bin/python --action_env=PYENV_ROOT --python_path=/nix/store/zdba9frlxj2ba8ca095win3nphsiiqhb-python3-3.10.8/bin/python --distinct_host_configuration=false --define PROTOBUF_INCLUDE_PATH=/nix/store/3vm2ya1b52rdvjbansv39zdv0qwmhdcz-protobuf-3.20.>
INFO: Found applicable config definition build:short_logs in file /build/source/.bazelrc: --output_filter=DONT_MATCH_ANYTHING
INFO: Found applicable config definition build:avx_posix in file /build/source/.bazelrc: --copt=-mavx --host_copt=-mavx
INFO: Found applicable config definition build:cuda in file /build/source/.bazelrc: --repo_env TF_NEED_CUDA=1 --action_env TF_CUDA_COMPUTE_CAPABILITIES=sm_35,sm_52,sm_60,sm_70,compute_80 --crosstool_top=@local_config_cuda//crosstool:toolchain --@local_config_cuda//:enable_cuda --@org_tensorflow//tensorflow/compiler/xla/python:enable_gpu=true --define=xla_python_enable_gpu=true
INFO: Found applicable config definition build:cuda in file /build/source/.jax_configure.bazelrc: --action_env TF_CUDA_COMPUTE_CAPABILITIES=sm_35,sm_50,sm_60,sm_70,sm_75,compute_80
INFO: Found applicable config definition build:mkl_open_source_only in file /build/source/.bazelrc: --define=tensorflow_mkldnn_contraction_kernel=1
INFO: Found applicable config definition build:linux in file /build/source/.bazelrc: --config=posix --copt=-Wno-unknown-warning-option --copt=-Wno-stringop-truncation --copt=-Wno-array-parameter
INFO: Found applicable config definition build:posix in file /build/source/.bazelrc: --copt=-fvisibility=hidden --copt=-Wno-sign-compare --cxxopt=-std=c++17 --host_cxxopt=-std=c++17
Loading:
Loading: 2 packages loaded
Analyzing: target //build:build_wheel (3 packages loaded, 0 targets configured)
Analyzing: target //build:build_wheel (46 packages loaded, 139 targets configured)
Analyzing: target //build:build_wheel (105 packages loaded, 855 targets configured)
Analyzing: target //build:build_wheel (215 packages loaded, 4629 targets configured)
Analyzing: target //build:build_wheel (235 packages loaded, 14022 targets configured)
INFO: Analyzed target //build:build_wheel (262 packages loaded, 15197 targets configured).
INFO: Found 1 target...
[0 / 60] [Prepa] Expanding template build/build_wheel ... (3 actions, 0 running)
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.devtools.build.lib.unsafe.StringUnsafe (file:/build/tmp/install/3cacfc6805558968a677264b92eb71de/A-server.jar) to constructor java.lang.String(byte[],byte)
WARNING: Please consider reporting this to the maintainers of com.google.devtools.build.lib.unsafe.StringUnsafe
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[25 / 96] Compiling tensorflow/compiler/xla/service/custom_call_status.cc; 1s local ... (16 actions, 15 running)
[47 / 109] Compiling absl/strings/internal/str_format/arg.cc; 3s local ... (16 actions, 15 running)
[78 / 133] Compiling absl/time/internal/cctz/src/time_zone_info.cc; 3s local ... (16 actions, 15 running)
[103 / 153] Compiling absl/synchronization/mutex.cc; 2s local ... (16 actions, 15 running)
[144 / 181] Compiling absl/strings/cord.cc; 3s local ... (16 actions, 15 running)
[240 / 1,622] Compiling jaxlib/ducc_fft.cc; 6s local ... (16 actions, 15 running)
[255 / 1,622] Compiling jaxlib/ducc_fft.cc; 9s local ... (16 actions, 15 running)
[270 / 1,622] Compiling jaxlib/ducc_fft_kernels.cc; 13s local ... (16 actions running)
[296 / 1,642] Compiling jaxlib/ducc_fft_kernels.cc; 17s local ... (16 actions, 15 running)
[362 / 1,690] Compiling jaxlib/ducc_fft_kernels.cc; 22s local ... (16 actions, 15 running)
[428 / 1,744] Compiling jaxlib/ducc_fft_kernels.cc; 29s local ... (16 actions, 15 running)
[518 / 1,816] Compiling jaxlib/ducc_fft_kernels.cc; 36s local ... (16 actions, 15 running)
ERROR: /build/output/external/local_config_cuda/cuda/BUILD:238:8: Executing genrule @local_config_cuda//cuda:cuda-include failed: (Exit 1): bash failed: error executing command /nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16/bin/bash -c ... (remaining 1 argument skipped)
cp: cannot stat '/nix/store/xzr4m07l1ix0820xbn6jzb5297l13w1v-cudatoolkit-11.8.0-merged/include/./include': No such file or directory
Target //build:build_wheel failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /build/output/external/local_config_cuda/cuda/BUILD:50:20 Middleman _middlemen/@local_Uconfig_Ucuda_S_Scuda_Ccuda_Uheaders-BazelCppSemantics_build_arch_k8-opt failed: (Exit 1): bash failed: error executing command /nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16/bin/bash -c ... (remaining 1 argument skipped)
INFO: Elapsed time: 57.694s, Critical Path: 37.63s
INFO: 547 processes: 159 internal, 388 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

**Comment by tartavull** [comment]
I have a very similar issue:

flake.nix
```
{
  description = "An awesome machine-learning project";
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.11";
    utils.url = "github:numtide/flake-utils";
    ml-pkgs.url = "github:nixvital/ml-pkgs";
    ml-pkgs.inputs.nixpkgs.follows = "nixpkgs";
    ml-pkgs.inputs.utils.follows = "utils";
  };

  outputs = { self, nixpkgs, ... }@inputs: {
    overlays.dev = nixpkgs.lib.composeManyExtensions [
      inputs.ml-pkgs.overlays.jax-family
      # })
    ];
  } // inputs.utils.lib.eachSystem [
    "x86_64-linux"
  ] (system:
    let pkgs = import nixpkgs {
          inherit system;
          config.allowUnfree = true;
          overlays = [ self.overlays.dev ];
        };
    in {
      devShells.default = let
        python-env = pkgs.python3.withPackages (pyPkgs: with pyPkgs; [
          jaxlibWithCuda11
        ]);
        name = "torch-basics";
      in pkgs.mkShell {
        inherit name;

        packages = [
          python-env
        ];
        shellHooks = let pythonIcon = "f3e2"; in ''
          export PS1="$(echo -e '\u${pythonIcon}') {\[$(tput sgr0)\]\[\033[38;5;228m\]\w\[$(tput sgr0)\]\[\033[38;5;15m\]} (${name}) \\$ \[$(tput sgr0)\]"
        '';
      };
    });
}
```

Output
```
> nix develop
error: builder for '/nix/store/z3wdc351wfr4s6qpzkpnb58xh6whyzc5-bazel-build-jaxlib-0.3.22.drv' failed with exit code 1;
       last 10 log lines:
       > ERROR: /build/output/external/local_config_cuda/cuda/BUILD:238:8: Executing genrule @local_config_cuda//cuda:cuda-include failed: (Exit 1): bash failed: error executing command /nix/store/kga2r02rmyxl14sg96nxbdhifq3rb8lc-bash-5.1-p16/bin/bash -c ... (remaining 1 argument skipped)
       > cp: cannot stat '/nix/store/0wg6hbjd4qjrj879wp1dnx70z06bjp6g-cudatoolkit-11.8.0-merged/include/./include': No such file or directory
       > Target //build:build_wheel failed to build
       > Use --verbose_failures to see the command lines of failed build steps.
       > ERROR: /build/output/external/org_tensorflow/tensorflow/core/BUILD:1708:23 Middleman _middlemen/@org_Utensorflow_S_Stensorflow_Score_Cstream_Uexecutor_Uheaders_Ulib-BazelCppSemantics_build_arch_k8-opt failed: (Exit 1): bash failed: error executing command /nix/store/kga2r02rmyxl14sg96nxbdhifq3rb8lc-bash-5.1-p16/bin/bash -c ... (remaining 1 argument skipped)
       > INFO: Elapsed time: 13.125s, Critical Path: 1.54s
       > INFO: 674 processes: 607 internal, 67 local.
       > FAILED: Build did NOT complete successfully
       > FAILED: Build did NOT complete successfully
       >
       For full logs, run 'nix log /nix/store/z3wdc351wfr4s6qpzkpnb58xh6whyzc5-bazel-build-jaxlib-0.3.22.drv'.
error: 1 dependencies of derivation '/nix/store/xn10jw0ba5sfv04njzhahxriwz8i8m9h-python3.10-jaxlib-0.3.22.drv' failed to build
error: 1 dependencies of derivation '/nix/store/8il99xqr7shb09jg8751mjjchlw7ahz9-python3-3.10.11-env.drv' failed to build
error: 1 dependencies of derivation '/nix/store/00vnm6q22mydhs4wf7lacd1a1plif6in-torch-basics-env.drv' failed to build
```

**Comment by tartavull** [comment]
The problem is clearly the line 
```cp: cannot stat '/nix/store/0wg6hbjd4qjrj879wp1dnx70z06bjp6g-cudatoolkit-11.8.0-merged/include/./include'```

`include` is concatenated twice and I verified that `'/nix/store/0wg6hbjd4qjrj879wp1dnx70z06bjp6g-cudatoolkit-11.8.0-merged/include` exists.

In Jax WORKSPACE file https://github.com/google/jax/blob/9a7f29a9fbb350dadbda42c89688f81ad4d59ab0/WORKSPACE#L8

Tensorflow is downloaded
```
http_archive(
    name = "org_tensorflow",
    sha256 = "9a7a7a87356bdeef5874fae135de380466482b593469035be3609a9cd2c153c4",
    strip_prefix = "tensorflow-cb946f223b9b3fa04efdbb7a0e6a9dabb22a7057",
    urls = [
        "https://github.com/tensorflow/tensorflow/archive/cb946f223b9b3fa04efdbb7a0e6a9dabb22a7057.tar.gz",
    ],
)
```

Tensorflow has the following file https://github.com/tensorflow/tensorflow/blob/58b37cf745c6c30d44878ddf5987c4283ed12c3c/third_party/gpus/cuda/BUILD.tpl
That define `cuda_headers`

```
cc_library(
    name = "cuda_headers",
    hdrs = glob([
        "**/*.h",
    ]),
    includes = [
        ".",
        "include",
    ],
    visibility = ["//visibility:public"],
)
```

Assuming that the includes list is what causing the extra "./include", I wonder if eliminating them would solve the issue.  This would require patching bazel to use a "local" copy of tensorflow.

**Comment by breakds** [comment]
Thanks for reporting the issue. I tried and verified that jax is not working with Cuda 11.8. @tartavull's analysis looks very reasonable. Perhaps a patch that removes the `"include"` line would work.

**Comment by breakds** [comment]
If `11.8` is not that important to you, I think you can downgrade to CUDA `11.7` and I verified that it works. This is basically equivalent to using the upstream `jaxlibWithCuda` from Nixpkgs.

**Comment by breakds** [comment]
> The problem is clearly the line `cp: cannot stat '/nix/store/0wg6hbjd4qjrj879wp1dnx70z06bjp6g-cudatoolkit-11.8.0-merged/include/./include'`
> 
> `include` is concatenated twice and I verified that `'/nix/store/0wg6hbjd4qjrj879wp1dnx70z06bjp6g-cudatoolkit-11.8.0-merged/include` exists.
> 
`cudatoolkit-11.7.0-merged` has the same structure. However, `jaxlib` can compile with it. I am not sure what has changed since 11.8.

**Comment by tartavull** [comment]
My assessment was wrong :) 

`/nix/store/0wg6hbjd4qjrj879wp1dnx70z06bjp6g-cudatoolkit-11.8.0-merged/include/include` existed but was a dangling symlink because nvidia changed the name of the architecture.

**Comment by breakds** [comment]
I see. Still, seems that the best fix would be to remove that `include` in bazel WORKSPACE?

**Comment by tartavull** [comment]
Closing this because the issue was resolved by nixpkgs

**Comment by breakds** [comment]
> Closing this because the issue was resolved by nixpkgs

I am currently pinning to `23.05` and it seemed that the fix hasn't been backported yet. Currently in order to use the fixed version you probably need to pin `nixpkgs` to `unstable`.

**Comment by breakds** [comment]
Finally this is fixed in `23.11`. See branch https://github.com/nixvital/ml-pkgs/tree/dev/23.11

Closing this now. Feel free re-open if it does not work.
